### ISSUE TITLE
New Proxy for interactive jobs

### ISSUE THREAD

**Issue by johrstrom** [OSC/ondemand#1306]
This is the epic ticket to replace our current proxy. While `mod_ood_proxy` has served us well for so many years, but now's the time start thinking about replacing it with something much more secure.

The attributes of such a solution should be:
* secure. No-one should be able to proxy to another user's interactive job.
* easy to install & maintain. It'd be nice to jump to some service mesh with etcd and all sorts of other components, but we can't really put that sort of administration burden on our administrators.
  * Compatibility is related to this. It should be able run at 90%+ of our sites.

**Comment by treydock** [comment]
Based on comments made at various conferences, not sure if this belongs in the code mentioned here or some other layer of OnDemand but it is useful for administrators or support staff to become other users. For OSC something like that is possible with Keycloak and impersonation but due to security issues people supporting OnDemand at OSC can't use that feature so some OnDemand-specific feature to support impersonation would be useful.  This could belong as the user mapping layer rather than the proxy layer.

**Comment by johrstrom** [comment]
Yea this will likely replace the lua code in `mod_ood_proxy` where it could take the authenticated user `tdockendorf` and actually proxy to `johrstrom`'s PUN or some other interactive session of `johrstrom`'s. 

So the comment/feature is in the right place, but I'm not sure when (if) that'll be available. One of the main benefits is to increase security and this adds a large attack vector, so I'd imagine that feature will be deferred for quite some time.

**Comment by johrstrom** [comment]
After thinking about this for sometime, I think I'm coming to the conclusion that we should likely be proxying through the PUN itself (the dashboard really). 

The issue is that there has to be some sort of `authorizing system` for managing and authorizing connections to the origins. The dashboard schedules the job and knows the connection information, so it has to publish this information to this `authorizing system`. 

Here's step by step of how this would work 

1. dashboard schedules the job 
2. dashboard publishes this information to the `authorizing system`
3. client request comes into apache (client wants to go to the origin)
4. apache has to communicate with and/or forward the request to the `authorizing system`
5. the `authorizing system` has to then make the determination to either forward the request or reject it.

If the dashboard itself is the `authorizing system` then that reduces a lot of complexity & maintenance on our side as developers. It would also reduce a lot of deployment complexity for administrators if they don't have to manage some external service like `envoy` or similar.

**Comment by jgoodson** [comment]
I was curious where the OSC team stood on proxy replacement. I've spent a lot of time thinking about and working on this issue: we run a modified proxy in production that accomplishes the security aspects of this but requires some external components that are not easily generalizable. I've had a lot of problems getting HTTPD to do what exactly I need without serious modifications or external components (what we use). I've gone two directions with prototypes to try to make this more transparent: one is as described above, retaining mod_ood_proxy or some variant of it and proxying through the PUN instead of directly to upstreams. I've also made some prototypes that drop HTTPD entirely in favor of some other reverse proxy, with the most successful prototype being one built around Caddy. 

Obviously the first is less invasive and a total replacement would be a serious, major-version-only change, if it is even feasible. I was curious if replacing HTTPD was something that had come up or been considered, or if sticking with HTTPD is important.

--

I do feel strongly that introducing more components or relying on things like service meshes or complex network namespace setups to make per-user overlays is undesirable. In my opinion, if the existing proxies (either HTTPD or the PUN) and the dashboard can be modified to accomplish this, that is ideal. In my opinion, the most straightforward way to accomplish this with the least code to maintain is to enforce mTLS use between the proxy and the upstream app server and to tie certificate validation to authorization, so only the correct authenticated user can get a proxy to connect to the upstream at all. This also has the advantage that there is no extra "system" that must be run, certificate validation provides authorization implicitly without needing some `authorizing system` service. 

This can be done with either a central proxy or per-user proxies, with different certificate setups for either. There are tradeoffs when it comes to introducing an extra hop (or two if you actually feed requests through the dashboard code) vs trying to get the central proxy do do this job. In either case, this does require changing how the jobs themselves run: either the apps must be configured to serve (m)TLS with the cert, or more likely the job would run a tunnel process like stunnel or ghostunnel to wrap the app and provide the TLS.

One issue I would point out about using the dashboard as the `authorization system` in general, when it comes to security, is that you can't really safely delegate a decision about whether a request should be allowed to a PUN, since you can't _really_ trust a dashboard. You really need to ensure the authorization comes from something controlled by the admins or owned by the user of the upstream/origin, not the requesting user. This means that if you aren't using a trusted central system you need a reliable way to know for sure who that "resource owner" is, which needs to be something secure and out of the request-user's control. Proxying through the PUN/dashboard doesn't directly accomplish this because you need a way to get authorization from the owner of the upstream, not the PUN/dashboard. Properly-configured TLS certificates can provide this guarantee, identifying the owner of the upstream with certainty and client authentication can prevent anyone but the authorized party (either the central proxy or the correct user's PUN depending on client CA config) from connecting to the app.

--

In the spirit of the above step-by-step, here are my two prototype designs described similarly:

PUN proxying
--------------
1. nginx_stage ensures per-user CA exists at PUN startup
2. dashboard schedules the job, provides access to the CA to the job
3. before app script runs, script wrapper signs a leaf certificate with the correct hostname with the per-user CA, sets up a TLS tunnel with client authentication requiring a cert signed by the per-user CA
4. application listens either on a UDS or on localhost on a local port that the TLS tunnel will connect to
5. client request comes in, apache matches request proxies request to the PUN
6. PUN proxies the request to the upstream over TLS (where that logic lives TBD), providing its client cert which is required for the handshake with the upstream to succeed due to mTLS. 

Central proxy
-------------
1. mod-portal-generator ensures "user certificate" and "infrastructure" CAs exists at generation time, without admin intervention unless configured
2. nginx_stage signs user-specific leaf cert at PUN startup with CN=username and with the "user certificate" CA
3. dashboard schedules the job and copies leaf cert and key to job directory
4. before app script runs, script wrapper sets up a TLS tunnel with client authentication requiring a cert signed by "infrastructure" CA
5. application listens either on a UDS or on localhost on a local port that the TLS tunnel will connect to
6. client request comes in, not-apache does what Apache does now, except that it connects to the upstream over TLS, providing its "infrastructure" CA-signed client cert, and sets SNI/expected hostname to the username so that cert validation requires the correct user cert (this last part is what I can't figure out how to make Apache mod_ssl do without the external resolution tricks we use, so if someone can figure that out this all gets much easier).

In either case a proxy is configured so that the outgoing request goes over a TLS connection expecting a certificate tied to the user's identity, either because it is signed by a per-user CA in the first case or because the CN is the username. In both cases the TLS tunnel server doing the TLS handshake will reject any connection that does not provide the appropriate client certificate.

In the first case, only that user's PUN should have access to the per-user CA so only it can connect. Another user can try to connect, but they won't have access to the correct client cert so the handshake will fail. In the second case, only the central proxy will have access to the "infrastructure" CA-signed client certificate so no one, even the user themselves, should be able to connect to the tunnel since only the OOD proxy has the correct certificate.

In either case, the underlying application is not protected by TLS unless it is directly serving TLS (which only some applications can do, Jupyter and websockify included), so a Unix domain socket is preferred if the application supports it since those can be protected with filesystem permissions. A localhost port wrapped with TLS still allows users with access to the compute node to directly connect to the underlying service, so that setup still requiring additional authentication to protect against that scenario. More exotic solutions are required to fully lock down applications that can only listen on TCP ports without mTLS on systems with shared nodes. My designs for this involve some configurable options in the app template that let you pick how the tunnel will connect to your particular app, allowing you to decide whether you will handle TLS or whether you'll serve on a TCP port or Unix socket.

The PUN proxy scenario can potentially be modified to use certificates in a way more akin to the central proxy scenario if the in-job certificate creation is unappealing enough. Having user's have access to the authentication secrets in some way is non-ideal, but unavoidable if the security is provided via the PUN/dashboard.

**Comment by johrstrom** [comment]
> curious if replacing HTTPD was something that had come up or been considered, or if sticking with HTTPD is important.

Maybe it's been suggested, but never taken seriously (at least not by me). I think stability is important, but it's also an issue of time (both ours the developers, but also the admin on the other side who's got to install and probably debug it at some point).

> Central proxy

For my money, I think the central proxy makes more sense as it requires less network hops. This will reduce not only latency, but also proxying bugs in any given layer. Also as you indicate, there are issues with PUN security (though admittedly I don't quite follow it).

That said - a PUN proxy may solve our `mod_substitute` problem, but I'd guess it's more error prone in general.

> In either case, the underlying application is not protected by TLS unless it is directly serving TLS (which only some applications can do, Jupyter and websockify included), so a Unix domain socket is preferred if the application supports it since those can be protected with filesystem permissions.

Yea origins need some work here. I feel like most can boot directly on https and have auth built-in (Jupyter for example). We're likely going to solve for these first, then tackle the harder ones later. In any case, Jupyter & Rstudio cover the majority of use cases, so just those 2 (that likely can boot on https and have built-in auth) will be a big win.

**Comment by jgoodson** [comment]
> Maybe it's been suggested, but never taken seriously (at least not by me). I think stability is important, but it's also an issue of time (both ours the developers, but also the admin on the other side who's got to install and probably debug it at some point).

>For my money, I think the central proxy makes more sense as it requires less network hops. This will reduce not only latency, but also proxying bugs in any given layer. Also as you indicate, there are issues with PUN security (though admittedly I don't quite follow it).

I'll continue to think about it, but I haven't worked out a way to accomplish both these goals with only Apache, especially with the TLS mechanisms, though a central proxy is my preference as well. The problem is that you need a one-size fits all solution for authorizing users for upstreams if you use a central proxy. If you use TLS, you need some way for the upstream certificate to only be valid for the requesting user, and I just can't seem to make it work with Apache, it doesn't have the hooks you need to modify the cert name validation process available with mod_proxy+mod_ssl. Hypothetically you could come up with some extra layer of authentication between the central proxy and the upstream, although it seems like a lot of dev work work and still adds an extra layer of network traffic that the proxy flow has to wait on and for bugs to crop up.

I've accomplished this with an another proxy replacing Apache entirely or beside Apache behind an HAProxy, or with other extra services beyond Apache, but Apache is too inflexible (without replacing mod_proxy or mod_ssl). I could imagine some extra layer of authorization for an upstream that you could implement as an Apache module. Maybe something like constructing JWTs for proxied request headers that get validated by some wrapper around the application before responding since we don't control the upstream servers unless there is a tunnel. You really need some upstream-side piece of this as best I can imagine, unless you have some third party that can somehow safely map upstreams to users.

In the Apache-replacement Caddy-only scenario, at least it is simple for admins because the setup/server software would be entirely provided by the package and wouldn't require or use OS-provided packages. It would certainly be a _different_ piece of software to troubleshoot, although likely not more difficult. It isn't really that much more code than mod_ood_proxy, though it would be all-new Golang. It would also mean any non-OSC-provided features would need to be for the other proxy and not Apache, so that would complicate the lives of admins who do a lot of customization depending on HTTPD features or modules. An Nginx solution would also work but you would need to either write a C API module or use an openresty build of Nginx and port mod_ood_proxy to their Lua API.

--

If you want to see an example of a secure PUN proxy, [this branch](https://github.com/jgoodson/ondemand/tree/integrated_pun_proxy) has a minimal, working if you set the OOD_PUN_PROXY var and have a functional DNS resolver, implementation. It manages its own per-user CA+certificates without admin involvement via `nginx_stage`, makes TLS certs accessible to jobs via the dashboard, and proxies through the PUN by having Apache `mod_ood_proxy` send requests to the PUN socket and pass along the host/URI information in headers. The PUN then uses the normal Nginx proxy_pass mechanism to reverse proxy, with SSL settings configured such that it can connect to a user's TLS upstreams with required client auth, mercifully not requiring an Nginx module. This does not require any additional software, components, or potentially even config (with some tweaks) if your app supports (m)TLS.

```
    location = /proxy {
      resolver <%= nginx_resolver %>;
      proxy_pass https://$http_x_ood_upstream$http_x_ood_req_uri;

      proxy_ssl_protocols TLSv1.2 TLSv1.3;
      proxy_ssl_verify on;
      proxy_ssl_name <%= user %>;
      proxy_ssl_server_name on;

      proxy_ssl_trusted_certificate <%= user_cert_ca %>;
      proxy_ssl_certificate <%= per_user_certificate %>;
      proxy_ssl_certificate_key <%= per_user_certificate_key %>;
      
      proxy_http_version 1.1;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection $connection_upgrade;
    }
```

The first section does the RP to the upstream. The second section turns on the SSL setup and the third provides the user-specific client/server cert setup. The final section is required for proxied websockets.

A key piece is `proxy_ssl_name <%= user %>;`. This makes Nginx validate the certificate using the user's name instead of the compute node hostname, allowing for user-specific certs. This isn't actually necessary with the PUN, because we could have the user mint their own job valid hostname-specific certificates to be validated by the PUN  with a per-user CA. I partly did this this way to illustrate that this is something you can do with Nginx that, as far as I can tell, is impossible with Apache but would enable the central proxy method with TLS authentication (since the central proxy could use the requesting user this way and validation would fail unless they matched the CN on the user's cert, which has their username in it and only they can use.

The PUN proxy method is definitely the least amount of new code and probably the easiest both for maintenance of OOD and administration of an install, at the cost of adding a layer to the actual request path and another place for bugs to crop up. If you integrate a tunnel with the job wrapper then admins don't have to do anything but install the tunnel software like they install websockify and port their apps appropriately. Also it is very possible to move more logic and request/response manipulation from mod_ood_proxy or the Apache directives to the PUN, but that was unnecessary for this purpose.

> That said - a PUN proxy may solve our mod_substitute problem, but I'd guess it's more error prone in general.

I'm personally iffy on the idea of manipulating response bodies, it is expensive and error prone when you aren't in control of the upstream software. You still have to have the requests hitting Apache use a URI prefix so PUN proxy doesn't solve anything on its own. You could presumably do the equivalent of mod_substitute in Nginx more easily than in OS-distributed Apache 2.4. The above branch also includes a commit which shows a method of bypassing this issue by introducing a third proxy method (`/dnode`) which accepts requests on the root URI via alternate, non-443 high ports, not requiring any URI prefix from the client, and dynamically routing this based on some server side state. This method is somewhat one-size fits all because it will work with, hypothetically, any web app.

I have some unanswered questions about how introducing another proxy layer interacts with some of the problematic use patterns we've observed or whether it adversely impacts performance. I can't imagine it being that bad, it is ultimately similar to a websocket terminal used via the PUN but with an extra SSL handshake in there. It would keep more PUNs alive longer since they wouldn't be idle while proxying. It is also a bit unfortunate that dynamic upstreams with Nginx require the use of a DNS resolver. It is a fundamental issue with using Nginx for this, though I expect most sites could have dnsmasq or systemd-resolved running on their OOD hosts.

> Yea origins need some work here. I feel like most can boot directly on https and have auth built-in (Jupyter for example). We're likely going to solve for these first, then tackle the harder ones later. In any case, Jupyter & Rstudio cover the majority of use cases, so just those 2 (that likely can boot on https and have built-in auth) will be a big win.

Jupyter is great, works with no issues and it is what I general use for testing all of this. RStudio Server is frustrating, because at least last I checked they restrict any SSL to the Server Pro (now Workbench) product. They really want you to buy Posit Workbench which has a very OOD-like feature set these days. If you want TLS-level authorization control as I've implemented, you also need client cert validation support which is somewhat less common than basic SSL support. If apps can listen on Unix sockets for a tunnel, that is just as good and a lot can do that. Most applications (things that use dynamically-linked libc for sockets so no static linking or Golang) can be forced to do so with a `LD_PRELOAD` shim. If all else fails and you have unprivileged network namespaces available you can accomplish it with those, whether with containers or more directly.

I could see in practice a mix of methods of protecting upstreams. If mTLS is available like with Jupyter, use it. If Unix sockets are usable, use them + stunnel. If they aren't, either a LD_PRELOAD shim (either OOD- or app author-provided) could be used or you just rely on application-level auth to protect it from people accessing the upstream, a modest improvement on the status quo since the OOD proxy won't enable it from the outside network anymore. It does put some burden on application authors to understand the different methods and determine what works for their app.

**Comment by johrstrom** [comment]
> After thinking about this for sometime, I think I'm coming to the conclusion that we should likely be proxying through the PUN itself (the dashboard really).

I'm now concerned about the throughput capabilities of the PUN. I.e., I don't believe the nginx is capable of such a throughput we'd need.

**Comment by treydock** [comment]
In my experience, NGINX is extremely well adapted at throughput.  It's one of the primary ingress handlers for Kubernetes where a single NGINX instance handles all ingress traffic, and I've done consulting work with large corporations that their ingress to their services was all NGINX.

Why the concern about the capabilities of NGINX?

**Comment by goodsonjr** [comment]
I am also curious about what the concerns are. I would probably place nginx somewhere around twice as fast as Apache mod_proxy + mpm_event, with less memory usage. Adding an additional, "unnecessary" hop is a concern, although I wouldn't expect it to seriously impact throughput, I've likely got more proxies in the request chain for OOD than most and proxy performance is not limiting. I also have nginx reverse proxies running that handle way more throughput than our ~300 concurrent user OOD instance, although not distributed across 300 independent nginx instances on one host... Per-user CAs require some extra plumbing, although having the flexibility of user processes minting their own certs is quite appealing.

--

That said, I have recently thought of a few other ideas, including one way to generalize our current production setup so it is suitable for other sites and has fully integrated configuration.

Our current setup, which is satisfactory for our use and that I have discussed before, uses synthetic hostnames containing usernames, matching backend certificates, and mTLS. Apache mod_proxy+mod_ssl encrypts the connection and directly prevents users from accessing other users' servers: Apache verifies that the backend certificate matches the user-specific synthetic hostname, and mTLS-capable backends reject anything that is not Apache/OOD infrastructure.

The stumbling blocks are:

- The system-level DNS resolver has to resolve synthetic hostnames to the correct IP. This requires site-specific setup since, unlike Nginx, you can't configure Apache hostname resolution for proxy host lookups.
- Our Slurm Prolog-based per-job certificate issuance is not portable to other schedulers. You need a trusted place to issue job-specific certs, and there are just not many scheduler-independent places to insert that with the current OOD architecture.

One solution is to use systemd to run Apache in a mount namespace and bind-mount `/etc/resolv.conf` so Apache’s normal resolution uses an OOD-specific resolver. We use `username.cn1234`-style synthetic hostnames now because that allows easy wildcard DNS entries in dnsmasq. We cannot use per-user wildcard certs with that hostname shape. However, CoreDNS can resolve synthetic hostnames of the form `cn1234.username.ood-proxy.internal` to the IP address of `cn1234` using regex rewrite rules, which lets us use both wildcard certs and fully dynamic synthetic hostname resolution.

This solves the first issue because OOD can provide a resolver used only by processes in that namespace. It can fall back to `/etc/hosts` and `/etc/resolv.conf` for non-synthetic hostnames (which will generally be needed for auth unless everything is going through a forward proxy) and does not interfere with the rest of the system. The main issue I see is that, with the current PUN startup method, PUNs and processes launched through `nginx_stage` would (I think) inherit that namespace. For most configurations this should be fine because CoreDNS would resolve ordinary names as usual, but a cookie-cutter CoreDNS config may not work everywhere.

This also addresses the second issue because the CoreDNS hostname shape allows user-specific wildcard certs. `*.username.ood-proxy.internal` is valid for `cn1234.username.ood-proxy.internal` and will validate only for the correct user namespace. We could use an OOD instance-owned CA to sign wildcard certs at PUN startup time and make them readable by the user. The cert and key could live on a shared filesystem or somewhere the dashboard can copy into session directories. The main caveat is lifetime: the cert needs to last at least as long as the max PUN lifetime plus however long a job may run after the PUN stops, unless the node-side setup can refresh or obtain certs independently. Also, if sites use a multi-segment hostname like `cn1234.hpc-site.local` it would need to be embedded into some configuration, maybe proxy+cert issuance, maybe coredns, to ensure the resolution works properly.

---

Another method is to accept that many applications need a node-side proxy/wrapper anyway to add TLS. In that case, the node-side proxy can become the authorization boundary instead of Apache.

For instance: mod_ood_proxy could attach a signed JWT specifying the authenticated OOD user, target session, and possibly other constraints. The node-side proxy would validate the token on each request and reject if the token does not authorize access to that backend. The request also needs to somehow be authenticated as coming from OOD infrastructure, either because the token is unforgeable and scoped tightly or because the node-side proxy also requires an OOD client certificate.

This is heavier because it adds request metadata and cryptographic validation per request rather than relying only on TLS handshake validation, though that needs some testing to determine the real impact. It adds an extra hop, but in many cases that hop is already necessary for applications that do not directly support (m)TLS, and it spends job resources rather than OOD host resources.

This addresses the "nefariously accessing another user's server" threat model if the node-side proxy validates the token correctly. It does not address the "user/Apache is tricked into connecting to a server actually run by someone else" threat model in the same way the hostname/TLS-based methods do, unless it is also combined with backend server identity validation somehow, TBD. It is simpler because it avoids the synthetic DNS stuff and much of the cert-management complexity, but lacks some of the security unless we come up with some not-complex backend validation mechanism.

**Comment by goodsonjr** [comment]
One more set of thoughts on the general topic of replacing `mod_ood_proxy` I feel I should throw out here:

Ultimately, if Apache remains the primary webserver and reverse proxy, `mod_lua` is a compelling entrypoint for the necessary custom behavior whether with `mod_ood_proxy` in its current Lua module form or otherwise. It is a simple, documented, stable public API that gives you quite a lot of the power of "full" Apache modules. It is pretty fast. You are relatively limited by what the embedded Lua interpreter can do and what parts of APR are exposed, but it is reasonably powerful. I have had great experiences creating Lua extension modules in Rust with [mlua](https://github.com/mlua-rs/mlua) if you need things you can't do with the embedded Lua. It is an easy thing to reach for when better performance or missing features are needed for new behaviors. In my experience, it has good performance and maintainability, with good safety properties as long as you are careful about Apache's threading/process model. Unlike the Nginx or HAProxy Lua implementations, mod_lua doesn't have any ability to integrate with the event loop, but oh well, we aren't trying to build a `mod_proxy` replacement with it.

Apache is quite powerful, extremely configurable, and has a huge library of modules to extend behavior. That said, we don't use most of that power or configuration, although it gives sites the ability to customize a lot without code changes. Compared to alternative reverse proxies like Nginx or newer Go runtime-based proxies like Caddy, actually extending and modifying its behavior is more difficult. The public APIs for Apache itself are extensive, but most functionality we need exists in modules like `mod_proxy` or `mod_ssl`, which themselves have much more limited public APIs. There is very little ability to manipulate their behavior beyond what you can do in config files, even from other Apache modules, which makes the sort of extremely dynamic behaviors utilizing more-niche features of `mod_proxy` and/or `mod_ssl` extremely difficult or impossible without forking/patching those modules directly, which is a big ask.

OOD's use of Apache isn't really that complex! Recreating all the behaviors we get out of Apache, with the notable exceptions of authentication and site admin customization, is pretty straightforward and reproducing it in another modern reverse proxy or proxy framework is quite accessible. Websocket handling, header/path rewriting, logging, etc are well- or even better-handled by other proxies. PUN management can benefit from closer integration or alternately being further abstracted out with something like `systemd-run` rather than direct `sudo` subprocess invocation. I've prototyped non-Apache setups (alongside the secure backend mechanisms) with a custom Caddy module and with a custom proxy written with Pingora. The resulting proxies are quite straightforward to build and package, with very few (or no) runtime dependencies. Replacing Apache entirely has a lot of advantages, but also some real risks. 

First risk: This loses all of the rich ecosystem of auth modules available to Apache. While there is support for many of the common auth mechanisms with these other proxies, they are not trivial for site admins to drop in and configure themselves as with Apache, and some are just not available or not as well supported. If we chose to say "you need to use OIDC or use some other method via an OIDC bridge like Dex" or some small number of supported methods, this is viable. Ship a good, flexible OIDC mechanism as the supported auth setup and its solid.

I _have_ tested out a scenario where I keep Apache around exclusively for auth, using it as an external forward auth provider that the reverse proxy checks in with (over a controlled channel) before setting up session cookies like `mod_auth_*` do internally. This works pretty well, although now you do have more components running. In this mode you'd keep Apache around but generate an extremely reduced config with the portal gen that essentially only has the auth setup and doesn't do any proxying or PUN management.

Second risk: When using Apache and `mod_ssl` as OOD does you benefit from being able to offload a lot of patching to the distro maintainers, but you lose some of that with some other proxies. If OpenSSL or Apache see a new vulnerability, much of the response is handled by distro maintainers shipping a patched version and sites updating their packages. Caddy has modules compiled in and ships as a static binary, which dramatically simplifies packaging (another plus), but does mean that if there was some vuln in the Go crypto/TLS stack (now with [FIPS mode](https://go.dev/doc/security/fips140)!) or HTTP implementation, it requires updating Go/Caddy upstream and shipping a new patch version. Similarly, with something like a Pingora-based proxy, the actual application code is owned by the OOD packages, although it can be linked against system OpenSSL.

I see actually moving off Apache entirely as feasible, but a big architectural change with some advantages in terms of simplicity, development/feature flexibility, better integration into a single, custom proxy application, but disadvantages with the loss of the Apache module ecosystem and free extensive site admin customization. Keeping Apache while meaningfully changing the proxy/TLS mechanism likely means modifying or replacing `mod_proxy` and/or `mod_ssl`, and that feels more daunting than moving to another proxy entirely, although for someone comfortable with old Apache C code that may be easier! For that reason, I've mostly focused on ways to keep Apache around but get it to do _the thing_ with as few extra moving parts as I can manage.

**Comment by johrstrom** [comment]
> Why the concern about the capabilities of NGINX?

OK really it's passenger. Passenger may not allow us to multi-thread. 

https://www.phusionpassenger.com/library/config/standalone/reference/#--concurrency-model-concurrency_model
